### PR TITLE
Add support for fetching album art (with unit tests)

### PR DESCRIPTION
**Pull request by glen3b**
This PR builds upon the work done by @auchter in #111, and implements support for the `albumart` MPD command. The core work was all his; I added unit tests, documentation, and some edge case tweaks. (If needed, I'm happy to squash commits, merge these changes into the original #111 branch, or otherwise.)

Protocol documentation references as per #111:

1. https://www.musicpd.org/doc/html/protocol.html?highlight=cover#the-music-database
2. https://www.musicpd.org/doc/html/protocol.html?highlight=cover#binary

### CODE REVIEW

**Comment by glen3b** [comment]
What fec298c is supposed to do is change the buffering settings on the two reader files to avoid problems. See comments on #111.

The way I've been testing (although reproducible buffer-related problems have been inconsistent at best) has been via an MPD proxy:
`socat TCP-LISTEN:6601 SYSTEM:"nc localhost 6600 | ./conduit_stdio --delay 5000000`
conduit_stdio is a program which passes through characters from stdin to stdout one-by-one with the given delay between calls in nanoseconds. I'm not at liberty to release its source but if you'd like to replicate this test it should be fairly simple to recreate.

The results have been inconclusive, on my end, as to whether fec298c actually helps with any problem. Given that I haven't been able to find a clear, definitive explanation of Python socket buffering behavior, this route might not lead to a certain bug-free path even in edge cases.

@Mic92 as a possible alternative, do you think it would be possible and/or a good idea to have client instances just keep one file in binary reader mode, and have `_read_line` do decoding? This way we don't have to care about nuances buffering behavior, but I don't know if there are any side effects that might occur by making that change.

File: mpd/base.py [hunk 1]
```diff
@@ -548,6 +551,77 @@ def _read_lines(self):
...
 
+    def _read_numbytes(self, numToRead):
+        retVal = bytearray()
+        while numToRead > 0:
+            readResult = self._rbfile.read(numToRead)
+            if len(readResult) == 0:
+                break
+            retVal.extend(readResult)
+            numToRead -= len(readResult)
+        return bytes(retVal)
```
Comment (Mic92) [inline mpd/base.py:562]: python naming style:

```suggestion
    def _read_bytes(self, amount):
        data = bytearray()
        while amount > 0:
            result = self._rbfile.read(amount)
            if len(result) == 0:
                break
            data.extend(result)
            amount -= len(result)
        return bytes(data)
```

File: mpd/base.py [hunk 2]
```diff
@@ -548,6 +551,77 @@ def _read_lines(self):
...
+            field, val = line.split(": ")
+            if field == "size":
+                size = int(val)
+            elif field == "binary":
+                chunk_size = int(val)
+        
+        if size is None:
+            size = chunk_size
+        
+        data = self._read_numbytes(chunk_size)
```
Comment (Mic92) [inline mpd/base.py:585]:
```suggestion
        data = self._read_bytes(chunk_size)
```

File: mpd/tests.py [hunk 3]
```diff
@@ -237,6 +247,147 @@ def test_readcomments(self):
...
+        self.assertRaises(mpd.ConnectionError,
+            lambda: self.client.albumart('a/full/path.mp3'))
+        
+        realWriteFunc.assert_called_with('albumart "a/full/path.mp3" "0"\n')
+        
+        self.assertIs(self.client._sock, None)
+
+    def test_binary_albumart_singlechunk_networkinterrupted(self):
+        # length 16
+        expectedBinary = b'\xA0\xA1\xA2\xA3\xA4\xA5\xA6\xA7\xA8\xA9\xAA\xAB\xAC\xAD\xAE\xAF'
```
Comment (Mic92) [inline mpd/tests.py:288]:
```suggestion
        expected_binary = b'\xA0\xA1\xA2\xA3\xA4\xA5\xA6\xA7\xA8\xA9\xAA\xAB\xAC\xAD\xAE\xAF'
```

File: mpd/tests.py [hunk 4]
```diff
@@ -237,6 +247,147 @@ def test_readcomments(self):
...
+        
+        self.MPDWillReturnBinary(b"binary: 16\n",
+            expectedBinary[0:9],
+            expectedBinary[9:14],
+            expectedBinary[14:16], b"\n", b"OK\n")
```
Comment (Mic92) [inline mpd/tests.py:293]:
```suggestion
            expected_binary[0:9],
            expected_binary[9:14],
            expected_binary[14:16], b"\n", b"OK\n")
```

File: mpd/tests.py [hunk 5]
```diff
@@ -237,6 +247,147 @@ def test_readcomments(self):
...
+        
+        realBinary = self.client.albumart('a/full/path.mp3')
```
Comment (Mic92) [inline mpd/tests.py:295]:
```suggestion
        real_binary = self.client.albumart('a/full/path.mp3')
```

File: mpd/tests.py [hunk 6]
```diff
@@ -237,6 +247,147 @@ def test_readcomments(self):
...
+        self.assertMPDReceived('albumart "a/full/path.mp3" "0"\n')
+        self.assertEqual(realBinary, expectedBinary)
```
Comment (Mic92) [inline mpd/tests.py:297]:
```suggestion
        self.assertEqual(real_binary, expected_binary)
```

File: mpd/tests.py [hunk 7]
```diff
@@ -237,6 +247,147 @@ def test_readcomments(self):
...
+    def test_binary_albumart_singlechunk_nosize(self):
+        # length: 16
+        expectedBinary = b'\x01\x02\x00\x03\x04\x00\xFF\x05\x07\x08\x0A\x0F\xF0\xA5\x00\x01'
+        
+        self.MPDWillReturnBinary(b"binary: 16\n",
+            expectedBinary, b"\n", b"OK\n")
+        
+        realBinary = self.client.albumart('a/full/path.mp3')
+        self.assertMPDReceived('albumart "a/full/path.mp3" "0"\n')
+        self.assertEqual(realBinary, expectedBinary)
```
Comment (Mic92) [inline mpd/tests.py:310]:
```suggestion
        expected_binary = b'\x01\x02\x00\x03\x04\x00\xFF\x05\x07\x08\x0A\x0F\xF0\xA5\x00\x01'
        
        self.MPDWillReturnBinary(b"binary: 16\n",
            expected_binary, b"\n", b"OK\n")
        
        real_binary = self.client.albumart('a/full/path.mp3')
        self.assertMPDReceived('albumart "a/full/path.mp3" "0"\n')
        self.assertEqual(real_binary, expected_binary)
```

File: mpd/tests.py [hunk 8]
```diff
@@ -237,6 +247,147 @@ def test_readcomments(self):
...
+    def test_binary_albumart_singlechunk_sizeheader(self):
+        # length: 16
+        expectedBinary = b'\x01\x02\x00\x03\x04\x00\xFF\x05\x07\x08\x0A\x0F\xF0\xA5\x00\x01'
+        
+        self.MPDWillReturnBinary(b"size: 16\n", b"binary: 16\n",
+            expectedBinary, b"\n", b"OK\n")
+        
+        realBinary = self.client.albumart('a/full/path.mp3')
+        self.assertMPDReceived('albumart "a/full/path.mp3" "0"\n')
+        self.assertEqual(realBinary, expectedBinary)
```
Comment (Mic92) [inline mpd/tests.py:325]:
```suggestion
        expected_binary = b'\x01\x02\x00\x03\x04\x00\xFF\x05\x07\x08\x0A\x0F\xF0\xA5\x00\x01'
        
        self.MPDWillReturnBinary(b"size: 16\n", b"binary: 16\n",
            expected_binary, b"\n", b"OK\n")
        
        real_binary = self.client.albumart('a/full/path.mp3')
        self.assertMPDReceived('albumart "a/full/path.mp3" "0"\n')
        self.assertEqual(real_binary, expected_binary)
```

File: mpd/tests.py [hunk 9]
```diff
@@ -237,6 +247,147 @@ def test_readcomments(self):
...
+        realBinary = self.client.albumart('a/full/path.mp3')
+        self.client._wfile.write.assert_has_calls([mock.call('albumart "a/full/path.mp3" "0"\n'),
+                                                mock.call('albumart "a/full/path.mp3" "16"\n'),
+                                                mock.call('albumart "a/full/path.mp3" "32"\n')])
+        self.assertEqual(realBinary, expectedBinary)
+
+        self.client._rbfile.readline.assert_has_calls([mock.call(),
+            mock.call(), mock.call(), mock.call(), mock.call(), mock.call()])
+        self.client._rbfile.read.assert_has_calls([mock.call(16), mock.call(1),
+            mock.call(16), mock.call(1), mock.call(16), mock.call(1)])
```
Comment (Mic92) [inline mpd/tests.py:353]:
```suggestion
        # length: 16 each
        expected_chunk1 = b'\x01\x02\x00\x03\x04\x00\xFF\x05\x07\x08\x0A\x0F\xF0\xA5\x00\x01'
        expected_chunk2 = b'\x0A\x0B\x0C\x0D\x0E\x0F\x10\x1F\x2F\x2D\x33\x0D\x00\x00\x11\x13'
        expected_chunk3 = b'\x99\x88\x77\xDD\xD0\xF0\x20\x70\x71\x17\x13\x31\xFF\xFF\xDD\xFF'
        expected_binary = expected_chunk1 + expected_chunk2 + expected_chunk3

        # 3 distinct commands expected
        self.MPDWillReturnBinary(b"size: 48\n", b"binary: 16\n",
            expected_chunk1, b"\n", b"OK\n", b"size: 48\n", b"binary: 16\n",
            expected_chunk2, b"\n", b"OK\n", b"size: 48\n", b"binary: 16\n",
            expected_chunk3, b"\n", b"OK\n")
        
        real_binary = self.client.albumart('a/full/path.mp3')
        self.client._wfile.write.assert_has_calls([mock.call('albumart "a/full/path.mp3" "0"\n'),
                                                mock.call('albumart "a/full/path.mp3" "16"\n'),
                                                mock.call('albumart "a/full/path.mp3" "32"\n')])
        self.assertEqual(real_binary, expected_binary)

        self.client._rbfile.readline.assert_has_calls([mock.call(),
            mock.call(), mock.call(), mock.call(), mock.call(), mock.call()])
        self.client._rbfile.read.assert_has_calls([mock.call(16), mock.call(1),
            mock.call(16), mock.call(1), mock.call(16), mock.call(1)])
```

File: mpd/tests.py [hunk 10]
```diff
@@ -237,6 +247,147 @@ def test_readcomments(self):
...
+        realBinary = self.client.albumart('a/full/path.mp3')
+        self.client._wfile.write.assert_has_calls([mock.call('albumart "a/full/path.mp3" "0"\n'),
+                                                mock.call('albumart "a/full/path.mp3" "17"\n'),
+                                                mock.call('albumart "a/full/path.mp3" "32"\n')])
+        self.assertEqual(realBinary, expectedBinary)
+
+        self.client._rbfile.readline.assert_has_calls([mock.call(),
+            mock.call(), mock.call(), mock.call(), mock.call(), mock.call()])
+        self.client._rbfile.read.assert_has_calls([mock.call(17), mock.call(1),
+            mock.call(15), mock.call(1), mock.call(1), mock.call(1)])
```
Comment (Mic92) [inline mpd/tests.py:377]:
```suggestion
    def test_binary_albumart_odd_multichunk(self):
        # lengths: 17, 15, 1
        expectedChunk1 = b'\x01\x02\x00\x03\x04\x00\xFF\x05\x07\x08\x0A\x0F\xF0\xA5\x00\x01\x13'
        expected_chunk2 = b'\x0A\x0B\x0C\x0D\x0E\x0F\x10\x1F\x2F\x2D\x33\x0D\x00\x00\x11'
        expected_chunk3 = b'\x99'
        expected_binary = expected_chunk1 + expected_chunk2 + expected_chunk3

        # 3 distinct commands expected
        self.MPDWillReturnBinary(b"size: 33\n", b"binary: 17\n",
            expected_chunk1, b"\n", b"OK\n", b"size: 33\n", b"binary: 15\n",
            expected_chunk2, b"\n", b"OK\n", b"size: 33\n", b"binary: 1\n",
            expected_chunk3, b"\n", b"OK\n")
        
        real_binary = self.client.albumart('a/full/path.mp3')
        self.client._wfile.write.assert_has_calls([mock.call('albumart "a/full/path.mp3" "0"\n'),
                                                mock.call('albumart "a/full/path.mp3" "17"\n'),
                                                mock.call('albumart "a/full/path.mp3" "32"\n')])
        self.assertEqual(real_binary, expected_binary)

        self.client._rbfile.readline.assert_has_calls([mock.call(),
            mock.call(), mock.call(), mock.call(), mock.call(), mock.call()])
        self.client._rbfile.read.assert_has_calls([mock.call(17), mock.call(1),
            mock.call(15), mock.call(1), mock.call(1), mock.call(1)])
```

File: mpd/tests.py [hunk 11]
```diff
@@ -237,6 +247,147 @@ def test_readcomments(self):
...
+    def test_binary_albumart_emptyresponse(self):
+        self.MPDWillReturnBinary(b"size: 0\n", b"binary: 0\n",
+            b"\n", b"OK\n")
+        
+        realBinary = self.client.albumart('a/full/path.mp3')
+        self.assertMPDReceived('albumart "a/full/path.mp3" "0"\n')
+        self.assertEqual(realBinary, b"")
+
+        self.client._rbfile.readline.assert_has_calls([mock.call(), mock.call()])
+        self.client._rbfile.read.assert_has_calls([mock.call(1)])
```
Comment (Mic92) [inline mpd/tests.py:389]:
```suggestion
    def test_binary_albumart_emptyresponse(self):
        self.MPDWillReturnBinary(b"size: 0\n", b"binary: 0\n",
            b"\n", b"OK\n")
        
        real_binary = self.client.albumart('a/full/path.mp3')
        self.assertMPDReceived('albumart "a/full/path.mp3" "0"\n')
        self.assertEqual(real_binary, b"")

        self.client._rbfile.readline.assert_has_calls([mock.call(), mock.call()])
        self.client._rbfile.read.assert_has_calls([mock.call(1)])
```

File: mpd/tests.py [hunk 12]
```diff
@@ -237,6 +247,147 @@ def test_readcomments(self):
...
+            b"\x00\x01\x02\x03")
+        
+        realWriteFunc = self.client._wfile.write
+
+        self.assertRaises(mpd.ConnectionError,
+            lambda: self.client.albumart('a/full/path.mp3'))
+        
+        realWriteFunc.assert_called_with('albumart "a/full/path.mp3" "0"\n')
+        
+        self.assertIs(self.client._sock, None)
```
Comment (Mic92) [inline mpd/tests.py:284]:
```suggestion
        self.MPDWillReturnBinary(b"size: 8\n", b"binary: 8\n",
            b"\x00\x01\x02\x03")
        
        real_write_func = self.client._wfile.write

        self.assertRaises(mpd.ConnectionError,
            lambda: self.client.albumart('a/full/path.mp3'))
        
        real_write_func.assert_called_with('albumart "a/full/path.mp3" "0"\n')
        
        self.assertIs(self.client._sock, None)
```

File: mpd/tests.py [hunk 13]
```diff
@@ -237,6 +247,147 @@ def test_readcomments(self):
...
+        
+        realWriteFunc = self.client._wfile.write
+
+        self.assertRaises(mpd.ConnectionError,
+            lambda: self.client.albumart('a/full/path.mp3'))
+        
+        realWriteFunc.assert_has_calls([mock.call('albumart "a/full/path.mp3" "0"\n'),
+            mock.call('albumart "a/full/path.mp3" "3"\n')])
+        
+        self.assertIs(self.client._sock, None)
```
Comment (Mic92) [inline mpd/tests.py:271]:
```suggestion
    def test_binary_albumart_disconnect_afterchunk(self):
        self.MPDWillReturnBinary(b"size: 17\n", b"binary: 3\n",
            b"\x00\x00\x00", b"\n", b"OK\n")
        
        real_write_func = self.client._wfile.write

        self.assertRaises(mpd.ConnectionError,
            lambda: self.client.albumart('a/full/path.mp3'))
        
        real_write_func.assert_has_calls([mock.call('albumart "a/full/path.mp3" "0"\n'),
            mock.call('albumart "a/full/path.mp3" "3"\n')])
        
        self.assertIs(self.client._sock, None)
```

**Comment by Mic92** [comment]
> What [fec298c](https://github.com/Mic92/python-mpd2/commit/fec298c8cd1fd5470d718fd1ea996728f83038e8) is supposed to do is change the buffering settings on the two reader files to avoid problems. See comments on #111.
> 
> The way I've been testing (although reproducible buffer-related problems have been inconsistent at best) has been via an MPD proxy:
> `socat TCP-LISTEN:6601 SYSTEM:"nc localhost 6600 | ./conduit_stdio --delay 5000000"`
> conduit_stdio is a program which passes through characters from stdin to stdout one-by-one with the given delay between calls in nanoseconds. I'm not at liberty to release its source but if you'd like to replicate this test it should be fairly simple to recreate.

Should be fine. I usually just tcpdump for stuff like this.

> 
> The results have been inconclusive, on my end, as to whether [fec298c](https://github.com/Mic92/python-mpd2/commit/fec298c8cd1fd5470d718fd1ea996728f83038e8) actually helps with any problem. Given that I haven't been able to find a clear, definitive explanation of Python socket buffering behavior, this route might not lead to a certain bug-free path even in edge cases.
> 
> @Mic92 as a possible alternative, do you think it would be possible and/or a good idea to have client instances just keep one file in binary reader mode, and have `_read_line` do decoding? This way we don't have to care about nuances buffering behavior, but I don't know if there are any side effects that might occur by making that change.

I am worried a separate connections just adds more complexity and edgecases in different areas. Maybe the connection failure test could be actually done using `socket.socketpair()` to create a fake server that does send the actual data instead.

**Comment by glen3b** [comment]
> > What [fec298c](https://github.com/Mic92/python-mpd2/commit/fec298c8cd1fd5470d718fd1ea996728f83038e8) is supposed to do is change the buffering settings on the two reader files to avoid problems. See comments on #111.
> > The way I've been testing (although reproducible buffer-related problems have been inconsistent at best) has been via an MPD proxy:
> > `socat TCP-LISTEN:6601 SYSTEM:"nc localhost 6600 | ./conduit_stdio --delay 5000000"`
> > conduit_stdio is a program which passes through characters from stdin to stdout one-by-one with the given delay between calls in nanoseconds. I'm not at liberty to release its source but if you'd like to replicate this test it should be fairly simple to recreate.
> 
> Should be fine. I usually just tcpdump for stuff like this.

I was trying to force the underlying syscalls not to return everything with one call. WireShark confirmed I was getting separate packets. In any case, the point is moot, as I didn't actually have problems when I reverted the commit and tested locally.

> > The results have been inconclusive, on my end, as to whether [fec298c](https://github.com/Mic92/python-mpd2/commit/fec298c8cd1fd5470d718fd1ea996728f83038e8) actually helps with any problem. Given that I haven't been able to find a clear, definitive explanation of Python socket buffering behavior, this route might not lead to a certain bug-free path even in edge cases.
> > @Mic92 as a possible alternative, do you think it would be possible and/or a good idea to have client instances just keep one file in binary reader mode, and have `_read_line` do decoding? This way we don't have to care about nuances buffering behavior, but I don't know if there are any side effects that might occur by making that change.
> 
> I am worried a separate connections just adds more complexity and edgecases in different areas. Maybe the connection failure test could be actually done using `socket.socketpair()` to create a fake server that does send the actual data instead.

I agree that the current situation is not ideal. To clarify, are you proposing that for the binary-related connections unit tests, instead of using `side_effect` on the read and readline calls, we actually set up a socket and have the unit tests write to the other end of it? I am open to that.

What I'm also wondering about is getting rid of the `_rfile` / `_rbfile` distinction entirely. That is, only keep one, probably `_rbfile`, and just have the client decode the returned bytestrings as needed. That way we can avoid the hacky buffer concerns. I'm wondering if you would be opposed to that, or if there's any pitfalls to doing this. (I don't know the codebase well.)

**Comment by Mic92** [comment]
> I agree that the current situation is not ideal. To clarify, are you proposing that for the binary-related connections unit tests, instead of using `side_effect` on the read and readline calls, we actually set up a socket and have the unit tests write to the other end of it? I am open to that.

Yes. The only issue could be that `socketpair()` might be not available on Windows. Also I don't care to much about it. A tcp-based solution might fix this.

> What I'm also wondering about is getting rid of the `_rfile` / `_rbfile` distinction entirely. That is, only keep one, probably `_rbfile`, and just have the client decode the returned bytestrings as needed. That way we can avoid the hacky buffer concerns. I'm wondering if you would be opposed to that, or if there's any pitfalls to doing this. (I don't know the codebase well.)

That sounds like a good solution. Decoding could be also done in the library itself since we return unicode strings everywhere.

**Comment by glen3b** [comment]
@Mic92 how's this?

I separated out the binary tests into a different class so I could initialize them differently. I also wrapped IOErrors in ConnectionErrors in `_read_binary`.

File: mpd/tests.py [hunk 14]
```diff
@@ -844,6 +683,169 @@ def test_command_list(self):
             ('xfade', '5')], sorted(res[5].items()))
 
 
+# MPD client tests which do not mock the socket, but rather replace it
+# with a real socket from a socket
+class TestMPDClientSocket(unittest.TestCase):
```
Comment (Mic92) [inline mpd/tests.py:688]: Thanks. This looks good! Can you make those tests being skipped on `windows`?

File: mpd/base.py [hunk 15]
```diff
@@ -523,7 +524,9 @@ def _write_command(self, command, args=[]):
         self._write_line(cmd)
 
     def _read_line(self):
-        line = self._rfile.readline()
+        line = self._rbfile.readline()
+        if not IS_PYTHON2:
```
Comment (Mic92) [inline mpd/base.py:528]:
```suggestion
        if self.use_unicode or not IS_PYTHON2:
```

File: mpd/tests.py (hunk 14 above)
Comment (glen3b) [reply mpd/tests.py:688]: Done, asterisk: I skip these tests specifically whenever `socketpair` is not available. From my brief testing, that means Python 2 on Windows. They pass with my Python 3.6 install on Windows.
